### PR TITLE
Documentation for ptw and tlb

### PR DESCRIPTION
**Pull request by midnighter95**
**Related issue**: 
Documentation

**Type of change**: bug report | feature request | other enhancement
other enhancement

**Impact**: no functional change | API addition (no impact on existing code) | API modification
no functional change

**Development Phase**: proposal |  implementation
implementation
**Release Notes**

Add documentation for TLB and PTW .
Unfinished work :  documentation for hypervisor mode IO signal , architecture details for the two-stage address translation

### CODE REVIEW

File: src/main/scala/rocket/PTW.scala [hunk 1]
```diff
@@ -26,11 +26,15 @@ class PTWReq(implicit p: Parameters) extends CoreBundle()(p) {
 class PTWResp(implicit p: Parameters) extends CoreBundle()(p) {
   val ae_ptw = Bool()
   val ae_final = Bool()
+  /**page fault  */
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:29]: please take care of the format issue.
```suggestion
  /** page fault  */
```

File: src/main/scala/rocket/PTW.scala [hunk 2]
```diff
@@ -26,11 +26,15 @@ class PTWReq(implicit p: Parameters) extends CoreBundle()(p) {
...
   val pf = Bool()
+  /**guest page fault  */
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:31]:
```suggestion
  /** guest page fault  */
```

File: src/main/scala/rocket/PTW.scala [hunk 3]
```diff
@@ -26,11 +26,15 @@ class PTWReq(implicit p: Parameters) extends CoreBundle()(p) {
...
   val gf = Bool()
+  //hypervisor wrx
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:33]: You need to read this: https://docs.scala-lang.org/overviews/scaladoc/for-library-authors.html

File: src/main/scala/rocket/PTW.scala [hunk 4]
```diff
@@ -26,11 +26,15 @@ class PTWReq(implicit p: Parameters) extends CoreBundle()(p) {
...
   val hr = Bool()
   val hw = Bool()
   val hx = Bool()
+  /**  return the target PTE */
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:37]: return to which component? where is the signal source, where are the signal sinks? This documentation doesn't give any additional information than the signal name: `PTE`.
```suggestion
  /** return the target PTE */
```

File: src/main/scala/rocket/PTW.scala [hunk 5]
```diff
@@ -39,6 +43,7 @@ class PTWResp(implicit p: Parameters) extends CoreBundle()(p) {
   val gpa_is_pte = Bool()
 }
 
+//send resp and CSRs info to TLB
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:46]:
```suggestion
// send resp and CSRs info to TLB
```

File: src/main/scala/rocket/PTW.scala [hunk 6]
```diff
@@ -74,7 +79,8 @@ class DatapathPTWIO(implicit p: Parameters) extends CoreBundle()(p)
   val customCSRs = coreParams.customCSRs.asInput
   val clock_enabled = Bool(OUTPUT)
 }
-
+/** Page table entry
+ *  @see RV-priv spec 4.3.1 for pgae table entry format*/
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:83]: Glad you have found the spec :)
```suggestion
/** Page table entry
  * @see RV-priv spec 4.3.1 for pgae table entry format*/
```

File: src/main/scala/rocket/PTW.scala [hunk 7]
```diff
@@ -87,8 +93,9 @@ class PTE(implicit p: Parameters) extends CoreBundle()(p) {
   val w = Bool()
   val r = Bool()
   val v = Bool()
-
+  //pointer to next level page table
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:96]: really? I only see a function here.

File: src/main/scala/rocket/PTW.scala [hunk 8]
```diff
@@ -87,8 +93,9 @@ class PTE(implicit p: Parameters) extends CoreBundle()(p) {
...
   def table(dummy: Int = 0) = v && !r && !w && !x && !d && !a && !u && reserved_for_future === 0
+  //leaf judge
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:98]: what does "judge" mean?

File: src/main/scala/rocket/PTW.scala [hunk 9]
```diff
@@ -113,21 +120,60 @@ class L2TLBEntry(nSets: Int)(implicit p: Parameters) extends CoreBundle()(p)
   val r = Bool()
 
 }
-
+/** PTW is the L2TLB, which perfrom page table walk for high level TLB.
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:123]:
```suggestion
/** PTW contains L2TLB, the PTW performs page table walk for high level TLB, and cache queries from L1 TLBs(I$, D$, RoCC)
```

File: src/main/scala/rocket/PTW.scala [hunk 10]
```diff
@@ -113,21 +120,60 @@ class L2TLBEntry(nSets: Int)(implicit p: Parameters) extends CoreBundle()(p)
...
+ *  It perform hierarchy page table query to mem for the desired leaf PTE and cache them in l2tlb.
+ *  Besides leaf PTEs,it also caches not-leaf PTEs in pte_cache to accerlerate the process
+ *
+ *  ==Structure==
+ *  - l2tlb : for leaf PTEs
+ *   - set-associative
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:129]: Is it configurable? which parameter decides the sets number?

**Review by sequencer — CHANGES_REQUESTED** [review]

File: src/main/scala/rocket/PTW.scala [hunk 11]
```diff
@@ -16,25 +16,36 @@ import freechips.rocketchip.util._
...
 
+/** PTE request from TLB to PTW
+  *
+  * TLB send a PTE request to PTW when TLB miss and it needs a page table walk
+  */
 class PTWReq(implicit p: Parameters) extends CoreBundle()(p) {
   val addr = UInt(width = vpnBits)
   val need_gpa = Bool()
   val vstage1 = Bool()
   val stage2 = Bool()
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:27]: what does these signals mean?
The interface signals should be well-documented:
from where, to where, the usage of this signal.

File: src/main/scala/rocket/PTW.scala [hunk 12]
```diff
@@ -16,25 +16,36 @@ import freechips.rocketchip.util._
...
 }
 
+/** PTE info from PTW to TLB
+  *
+  * containing: target PTE, exceptions, two-satge tanslation info */
 class PTWResp(implicit p: Parameters) extends CoreBundle()(p) {
   val ae_ptw = Bool()
   val ae_final = Bool()
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:35]: `ae` means access fault, I think those signals should be well documented as well.

File: src/main/scala/rocket/PTW.scala [hunk 13]
```diff
@@ -16,25 +16,36 @@ import freechips.rocketchip.util._
...
   val gf = Bool()
-  //hypervisor wrx
+  /** hypervisor read*/
   val hr = Bool()
+  /** hypervisor write*/
   val hw = Bool()
+  /** hypervisor execute*/
   val hx = Bool()
-  /**  return the target PTE */
+  /** target leaf PTE for a page table walk
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:46]: I don't know what does this mean.

File: src/main/scala/rocket/PTW.scala [hunk 14]
```diff
@@ -150,19 +221,25 @@ class PTW(n: Int)(implicit edge: TLEdgeOut, p: Parameters) extends CoreModule()(
   val resp_hx = Reg(Bool())
   val resp_fragmented_superpage = Reg(Bool())
 
+  /** tlb request */
   val r_req = Reg(new PTWReq)
+  /** current selected way in arbitor */
   val r_req_dest = Reg(Bits())
+    //to respond to L1TLB : l2_hit
+    //to construct mem.req.addr
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:229]: format....

File: src/main/scala/rocket/PTW.scala [hunk 15]
```diff
@@ -150,19 +221,25 @@ class PTW(n: Int)(implicit edge: TLEdgeOut, p: Parameters) extends CoreModule()(
...
   val r_pte = Reg(new PTE)
   val r_hgatp = Reg(new PTBR)
-
+  //2-stage pageLevel
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:232]:
```suggestion
  // 2-stage pageLevel
```

File: src/main/scala/rocket/PTW.scala [hunk 16]
```diff
@@ -199,14 +278,18 @@ class PTW(n: Int)(implicit edge: TLEdgeOut, p: Parameters) extends CoreModule()(
     val vpn_idx  = vpn_idxs(count) & mask
     val raw_pte_addr = ((r_pte.ppn << pgLevelBits) | vpn_idx) << log2Ceil(xLen / 8)
     val size = if (usingHypervisor) vaddrBits else paddrBits
+    //use r_pte.ppn as page table base address
+    //use vpn slice as offset
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:282]:
```suggestion
    //  use r_pte.ppn as page table base address
    // use vpn slice as offset
```

File: src/main/scala/rocket/PTW.scala [hunk 17]
```diff
@@ -199,14 +278,18 @@ class PTW(n: Int)(implicit edge: TLEdgeOut, p: Parameters) extends CoreModule()(
...
     raw_pte_addr.apply(size.min(raw_pte_addr.getWidth) - 1, 0)
   }
+  /** pte_cache input addr*/
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:285]:
```suggestion
  /** pte_cache input addr */
```

File: src/main/scala/rocket/PTW.scala [hunk 18]
```diff
@@ -199,14 +278,18 @@ class PTW(n: Int)(implicit edge: TLEdgeOut, p: Parameters) extends CoreModule()(
...
   val pte_cache_addr = if (!usingHypervisor) pte_addr else {
     val vpn_idxs = (0 until pgLevels-1).map(i => (aux_pte.ppn >> (pgLevels-i-1)*pgLevelBits)(pgLevelBits-1,0))
     val vpn_idx = vpn_idxs(count)
     val raw_pte_cache_addr = Cat(r_pte.ppn, vpn_idx) << log2Ceil(xLen/8)
     raw_pte_cache_addr(vaddrBits.min(raw_pte_cache_addr.getWidth)-1, 0)
   }
+  /** stage2_pte_cache input addr*/
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:292]:
```suggestion
  /** stage2_pte_cache input addr */
```

**Review by sequencer — COMMENTED** [review]
take care of the format issues....

File: src/main/scala/rocket/PTW.scala (hunk 13 above)
Comment (midnighter95) [reply src/main/scala/rocket/PTW.scala:46]: It should be the PTE to refill L1TLB from L2TLB. I have revised it.

File: src/main/scala/rocket/PTW.scala [hunk 19]
```diff
@@ -52,14 +76,20 @@ class TLBPTWIO(implicit p: Parameters) extends CoreBundle()(p)
...
-
+/** PTW performance statistics to Core */
 class PTWPerfEvents extends Bundle {
   val l2miss = Bool()
   val l2hit = Bool()
   val pte_miss = Bool()
   val pte_hit = Bool()
 }
 
+/** IO between PTW and Core
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:87]: Datapath IO

File: src/main/scala/rocket/PTW.scala [hunk 20]
```diff
@@ -74,7 +104,9 @@ class DatapathPTWIO(implicit p: Parameters) extends CoreBundle()(p)
   val customCSRs = coreParams.customCSRs.asInput
   val clock_enabled = Bool(OUTPUT)
 }
-
+/** Page table entry
+  * @see RV-priv spec 4.3.1 for pgae table entry format
+  */
 class PTE(implicit p: Parameters) extends CoreBundle()(p) {
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:110]: all these `val`s should be clearly documented.

File: src/main/scala/rocket/PTW.scala [hunk 21]
```diff
@@ -87,8 +119,9 @@ class PTE(implicit p: Parameters) extends CoreBundle()(p) {
   val w = Bool()
   val r = Bool()
   val v = Bool()
-
+  /** return true if find a pointer to next level page table */
   def table(dummy: Int = 0) = v && !r && !w && !x && !d && !a && !u && reserved_for_future === 0
+  /** return true if find a leaf PTE */
   def leaf(dummy: Int = 0) = v && (r || (x && !w)) && a
   def ur(dummy: Int = 0) = sr() && u
   def uw(dummy: Int = 0) = sw() && u
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:127]: these two functions?

File: src/main/scala/rocket/PTW.scala [hunk 22]
```diff
@@ -140,6 +212,7 @@ class PTW(n: Int)(implicit edge: TLEdgeOut, p: Parameters) extends CoreModule()(
   withClock (gated_clock) { // entering gated-clock domain
 
   val invalidated = Reg(Bool())
+  /** current PTE level */
```
Comment (sequencer) [inline src/main/scala/rocket/PTW.scala:215]: count from 0 or max level?

File: src/main/scala/rocket/TLB.scala [hunk 23]
```diff
@@ -64,21 +94,39 @@ class TLBResp(implicit p: Parameters) extends CoreBundle()(p) {
...
+  /** hypervisor excute */
   val hx = Bool()
+  /** hypervisor read */
   val hr = Bool()
+  /** prot_w */
   val pw = Bool()
+  /** prot_x */
   val px = Bool()
+  /** prot_r */
   val pr = Bool()
```
Comment (sequencer) [inline src/main/scala/rocket/TLB.scala:128]: Good job.

File: src/main/scala/rocket/PTW.scala (hunk 11 above)
Comment (midnighter95) [reply src/main/scala/rocket/PTW.scala:27]: Work in IO signals in hypervisor mode has been marked unfinished.

**Comment by jerryz123** [comment]
To reduce increasing LoC too much, I would prefer is short single line comments were attached with `//` to the end of the line.
Screen real estate is precious.

**Review by sequencer — APPROVED** [review]
LGTM for a first step documentation for now. Merge this for not blocking the chisel3 porting missing.

**Pull request merged**